.github/workflows/qemu.yaml

```
@@ -0,0 +1,72 @@
+name: QEMU environment
+
+on:
+  push:
+    branches:
+      - main
+  pull_request:
+    branches:
+      - main
+  release:
+    types:
+      - published
+
+jobs:
+  build-linux-arm64:
+    runs-on: ubuntu-22.04
+    steps:
+      - name: Install VM dependencies
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y qemu-system-arm qemu-efi-aarch64 qemu-utils cloud-image-utils wget
+      - name: Provision VM
+        run: |
+          wget https://cloud-images.ubuntu.com/jammy/current/jammy-server-cloudimg-arm64.img
+
+          qemu-img create -f qcow2 -F qcow2 \
+            -b jammy-server-cloudimg-arm64.img \
+            jammy-snapshot.qcow2 200G
+
+          cat >"user-data" << EOF
+          #cloud-config
+          password: passw0rd
+          chpasswd: { expire: False }
+          ssh_pwauth: True
+          EOF
+          cloud-localds user-data.img user-data
+
+          cp /usr/share/AAVMF/AAVMF_CODE.fd .
+          cp /usr/share/AAVMF/AAVMF_VARS.fd .
+
+          randmac=$(openssl rand -hex 6 | sed 's/\(..\)/\1:/g; s/:$//')
+
+          qemu-system-aarch64 \
+            -m 2048 \
+            -cpu max -smp 4 \
+            -machine virt,gic-version=max \
+            -display none \
+            -drive if=pflash,format=raw,readonly=on,file=AAVMF_CODE.fd \
+            -drive if=pflash,format=raw,file=AAVMF_VARS.fd \
+            -drive if=none,file=jammy-server-cloudimg-arm64.img,id=hd0 \
+            -drive file=user-data.img,format=raw \
+            -device virtio-blk-device,drive=hd0 \
```

```suggestion
            -drive if=none,file=jammy-server-cloudimg-arm64.img,id=hd0,cache=unsafe \
            -device virtio-blk-device,drive=hd0 \
            -drive file=user-data.img,format=raw,media=cdrom,id=cd0 \
            -device virtio-blk-device,drive=cd0 \
```
I think you're missing a -device to map the cdrom drive

---

Formatting 'jammy-snapshot.qcow2', fmt=qcow2 cluster_size=65536 extended_l2=off compression_type=zlib size=214748364800 backing_file=jammy-server-cloudimg-arm64.img backing_fmt=qcow2 lazy_refcounts=off refcount_bits=16
qemu-system-aarch64: -drive file=user-data.img,format=raw,media=cdrom,id=cd0: Drive 'cd0' is already in use because it has been automatically connected to another device (did you need 'if=none' in the drive options?)
Error: Process completed with exit code 1.